### PR TITLE
MEN-3996 Sessions and shell starting/stopping support.

### PR DESCRIPTION
**Pull request by merlin-northern**
MEN-3996 Sessions and shell starting/stopping support.

ChangeLog:none
Signed-off-by: Peter Grzybowski <peter@northern.tech>

### CODE REVIEW

**Comment by merlin-northern** [comment]
@tranchitella : tested manually, completely lacks tests in code, and some cleaning, but should be enough to review the basic idea and test with deviceconnect and ui patches.

File: app/daemon.go [hunk 1]
```diff
@@ -44,9 +53,13 @@ func NewDaemon(config *configuration.MenderShellConfig) *MenderShellDaemon {
 		username:         config.User,
 		shell:            config.ShellCommand,
 		serverUrl:        config.ServerURL,
+		skipVerify:       config.SkipVerify,
 		deviceConnectUrl: configuration.DefaultDeviceConnectPath,
+		terminalString:   configuration.DefaultTerminalString,
 		terminalWidth:    config.Terminal.Width,
 		terminalHeight:   config.Terminal.Height,
+		shellsSpawned:    0,
+		debug:            true,
```
Comment (tranchitella) [inline app/daemon.go:62]: I suppose this should be set to true from a (global?) CLI flag managed by `urfave/cli/v2`

File: session/session.go [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+		p.Signal(syscall.SIGKILL)
+		time.Sleep(time.Second)
+		done := make(chan error, 1)
+		go func() {
+			done <- s.command.Wait()
+		}()
+		select {
+		case err := <-done:
+			if err != nil {
+				//log.Errorf("failed to wait for the shell process to exit: %s", err.Error())
```
Comment (tranchitella) [inline session/session.go:284]: Why is this log commented?

**Review by tranchitella — APPROVED** [review]
I have a few comments, that will need follow-ups.
To be honest, I haven't reviewed the tests.
In my opinion, we can merge as it is and take care of the follow-ups with additional PRs.

File: session/session.go (hunk 2 above)
Comment (merlin-northern) [reply session/session.go:284]: no idea. uncommented. good catch.

File: session/session.go [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+	}
+
+	s.pseudoTTY.Close()
+	p, _ := os.FindProcess(s.shellPid)
+	if p != nil {
+		p.Signal(syscall.SIGINT)
+		time.Sleep(time.Second)
+		p.Signal(syscall.SIGTERM)
+		time.Sleep(time.Second)
+		time.Sleep(time.Second)
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply session/session.go:274]: copy paste. fixed.

File: session/session.go [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+	s.pseudoTTY = pseudoTTY
+	s.command = cmd
+	return nil
+}
+
+func (s *MenderShellSession) GetId() string {
+	return s.id
+}
+
+func (s *MenderShellSession) IsExpired() bool {
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply session/session.go:238]: https://tracker.mender.io/browse/MEN-4162

File: session/session.go [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+	ErrSessionShellAlreadyRunning         = errors.New("shell is already running")
+	ErrSessionShellNotRunning             = errors.New("shell is not running")
+	ErrSessionShellStillRunning           = errors.New("shell is still running")
+	ErrSessionShellTooManySessionsPerUser = errors.New("user has too many open sessions")
+	ErrSessionNotFound                    = errors.New("session not found")
+	ErrSessionTooManyShellsAlreadyRunning = errors.New("too many shells spawned")
+)
+
+var (
+	defaultSessionExpiredTimeout = 512 * time.Second
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply session/session.go:60]: bumped to 1024s

File: app/daemon.go (hunk 1 above)
Comment (merlin-northern) [reply app/daemon.go:62]: yes. https://tracker.mender.io/browse/MEN-4163

**Pull request merged**